compiler/src/dotty/tools/dotc/ast/DesugarEnums.scala

```diff
@@ -84,35 +92,55 @@ object DesugarEnums {
   private def valuesDot(name: PreName)(implicit src: SourceFile) =
     Select(Ident(nme.DOLLAR_VALUES), name.toTermName)
 
-  private def registerCall(using Context): Tree =
-    Apply(valuesDot("register"), This(EmptyTypeIdent) :: Nil)
+  private def ArrayLiteral(values: List[Tree], tpt: Tree)(using Context): Tree =
+    val clazzOf = TypeApply(ref(defn.Predef_classOf.termRef), tpt :: Nil)
+    val ctag    = Apply(TypeApply(ref(defn.ClassTagModule_apply.termRef), tpt :: Nil), clazzOf :: Nil)
+    val apply   = Select(ref(defn.ArrayModule.termRef), nme.apply)
+    Apply(Apply(TypeApply(apply, tpt :: Nil), values), ctag :: Nil)
 
-  /**  The following lists of definitions for an enum type E:
+  /**  The following lists of definitions for an enum type E and known value cases e_0, ..., e_n:
    *
-   *   private val $values = new EnumValues[E]
-   *   def values = $values.values.toArray
+   *   private val $values = Array[E](e_0,...,e_n)(ClassTag[E](classOf[E]))
+   *   @annotation.threadUnsafe private lazy val $valuesReverse =
+   *     scala.runtime.ScalaRuntime.wrapRefArray($values).map((x_0: E) => (x_0.enumLabel, x_0)).toMap
+   *   def values = $values.clone
    *   def valueOf($name: String) =
-   *     try $values.fromName($name) catch
+   *     try $valuesReverse($name) catch
```

Comment:
Since `$valuesReverse` is now statically known, and since it is only used here, I believe an even better strategy would be to emit a hard-code `match` here, i.e.,
```scala
def valueOf($name: String) =
  $name match
    case l_0 => e_0
    ...
    case l_n => e_n
    case _ => throw new IllegalArgumentException("enum case not found: " + $name)
```
where `l_i` are the hard-coded labels as string literals.

Currently, dotc generates "bad" code for such a `match`, because it will do a linear scan. However, the latest Scala 2.12 and 2.13 compilers already emit efficient code for `match`es that have only string literal cases. They emit a `switch` on the `hashCode` of the string, which is basically optimal (better than a lookup in a dynamic map). Since we'll have to improve the pattern matcher and back-end to support that, not to stay behind 2.13, I think it would be fine to emit a `match` even now.

This would also get rid of the synthetic dependency on Scala's `Map`.

Reply:
We could also decide to defer that to when we actually implement the improvements to the back-end.

Reply:
I am happy to change to a match, i did the map because i benchmarked against match

Reply:
there is a sketch here now but will try to merge the global optimisation for string matches before this PR.